BUG: Ensure Index.astype('category') returns a CategoricalIndex

jschendel opened the pull request:
- [X] closes #18630
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Notes:
  - `MultiIndex.astype('category')` raises per @TomAugspurger's comment in the issue.
  - `IntervalIndex.astype('category')` return a `Categorical` with `ordered=True` instead of `CategoricalIndex`, since it looks like someone previously intentionally implemented it this way.  I don't immediately see a reason why, but left it as is.  Would be straightforward to make this consistent and return a `CategoricalIndex`.
  -  All other types of index should return a `CategoricalIndex`.

jschendel committed:
BUG: Ensure Index.astype('category') returns a CategoricalIndex

jschendel commented on pandas/core/dtypes/common.py:
```diff
@@ -1934,7 +1934,7 @@ def pandas_dtype(dtype):
                 pass
 
-        elif dtype.startswith('interval[') or dtype.startswith('Interval['):
+        elif dtype.startswith('interval') or dtype.startswith('Interval'):
```
Changed this because switching to `pandas_dtype` caused a test to break since it was passing `'interval'` as the dtype, which appears to be valid:
```python
In [1]: from pandas.core.dtypes.common import is_interval_dtype

In [2]: is_interval_dtype('interval')
Out[2]: True
```

jschendel commented on pandas/core/indexes/multi.py:
```diff
@@ -2715,7 +2716,7 @@ def difference(self, other):
-        if not is_object_dtype(np.dtype(dtype)):
+        if not is_object_dtype(pandas_dtype(dtype)):
             raise TypeError('Setting %s dtype to anything other than object '
                             'is not supported' % self.__class__)
```
Made this change since `np.dtype` doesn't recognize `'category'`, and would raise a different error message than the one specified here; `pandas_dtype` ensures that this error message will be raised.

jreback replied:
ahh yes, we should have very very limited ``np.dtype`` calls anywhere; ``pandas_dtype`` is the general version

jorisvandenbossche commented:
> IntervalIndex.astype('category') return a Categorical with ordered=True instead of CategoricalIndex, since it looks like someone previously intentionally implemented it this way.

I think it should be CategoricalIndex

jreback commented on pandas/core/indexes/category.py:
```diff
@@ -341,9 +342,13 @@ def __array__(self, dtype=None):
         if is_interval_dtype(dtype):
             from pandas import IntervalIndex
             return IntervalIndex.from_intervals(np.array(self))
+        elif is_categorical_dtype(dtype) and (dtype == self.dtype):
```
you should use ``dtype.equals(self.dtype)`` here I think

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -1058,3 +1059,20 @@ def test_putmask_with_wrong_mask(self):
         with pytest.raises(ValueError):
             index.putmask('foo', 1)
+
+    def test_astype_category(self):
```
can you pass a ``CategoricalDtype`` here as well (with and w/o ordered) and make this parameterized

jreback commented on pandas/tests/indexes/test_interval.py:
```diff
@@ -362,8 +363,15 @@ def test_astype(self, closed):
         assert result.equals(idx)
 
-        result = idx.astype('category')
+    def test_astype_category(self, closed):
```
same

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -554,6 +555,15 @@ def test_astype(self):
         with tm.assert_raises_regex(TypeError, "^Setting.*dtype.*object"):
             self.index.astype(np.dtype(int))
 
+    def test_astype_category(self):
```
same

jschendel commented:
Question regarding how this should behave with a `CategoricalIndex`.

Setup:
```python
In [3]: ci = pd.CategoricalIndex(list('abca'))

In [4]: new_dtype = CategoricalDtype(ordered=True)

In [5]: new_dtype
Out[5]: CategoricalDtype(categories=None, ordered=True)

In [6]: ci.dtype
Out[6]: CategoricalDtype(categories=['a', 'b', 'c'], ordered=False)
```

Then `ci.dtype` and `new_dtype` are considered equal despite `ordered` being different:
```python
In [7]: ci.dtype == new_dtype
Out[7]: True
```
Which appears to be intentional, per the comment prior to the implementation:

https://github.com/pandas-dev/pandas/blob/9629fef912dbeca5154e9cf58970bad3e02657f4/pandas/core/dtypes/dtypes.py#L216-L222

In this case, should doing `ci.astype(new_dtype)` change `ordered` to `True`?
  - From a user standpoint I'd expect this to change `ordered` to `True`, as this would logically be the user's intention.
  - From a codebase consistency standpoint, it seems like since the two dtypes are equal, doing an `astype` shouldn't change anything

jschendel committed:
review updates

jschendel commented:
Made review related updates and a few additional fixes/changes:
  - `IntervalIndex.astype('category')` has been changed to return a `CategoricalIndex`
  - Modified `MultiIndex.astype('category')` to raise a `NotImplementedError` stating that >1 ndim categoricals aren't currently supported
      - Mirrors a similar error message raised by `DataFrame.astype('category')`.

Also modified the `CategoricalIndex.astype` behavior to address my question above.  I chose the first bulleted option.  The logic I implemented is that if a `CategoricalDtype` with `categories` or `ordered` being `None` is passed, `None` is replaced by the corresponding value from the `CategoricalIndex`'s dtype.  Basically `None` corresponds to "don't change this attribute".

The only slightly counter-intuitive thing is that `pandas_dtype('category')` returns a `CategoricalDtype` with `ordered=False`, meaning that if you have an ordered `CategoricalIndex` and do `.astype('category')`  it would keep the same categories but switch ordered to `False`.  Should be straightforward to make this not change anything if that'd be preferred.

jschendel commented on pandas/core/indexes/category.py:
```diff
@@ -341,9 +341,26 @@ def __array__(self, dtype=None):
+            dtype = CategoricalDtype(new_categories, new_ordered)
+
+            # fastpath if dtypes are equal
+            if dtype == self.dtype:
```
Note that `dtype.equals(self.dtype)` raises `AttributeError: 'CategoricalDtype' object has no attribute 'equals'`.

The only other thing I found for dtype comparison is `is_dtype_equal`, but under the hood that basically just does `==` within a `try`/`except` to catch if the dtypes aren't comparable.  We're within `elif is_categorical_dtype(dtype)` here though, so `is_dtype_equal` seem superfluous, but could still use it if preferred.

jorisvandenbossche commented:
> should doing ci.astype(new_dtype) change ordered to True?

I agree that from a user point of view this should change the ordered attribute.

> From a codebase consistency standpoint, it seems like since the two dtypes are equal, doing an astype shouldn't change anything

I am wondering if it would not be possible to change this. It is clear that we need to keep `cat_dtype == 'category'` to be True. But that is already handled a few lines above the snippet you showed. 
So I don't fully understand the comment why it is needed that all CDT(None, ordered=True/False) needs to be equal regardless of the ordered attribute.

cc @TomAugspurger 

>  if you have an ordered CategoricalIndex and do .astype('category') it would keep the same categories but switch ordered to False. Should be straightforward to make this not change anything if that'd be preferred.

That doesn't sound fully as the desired behaviour .. although I am not fully sure :-)

TomAugspurger commented:
> >  if you have an ordered CategoricalIndex and do .astype('category') it would keep the same categories but switch ordered to False. Should be straightforward to make this not change anything if that'd be preferred.

> That doesn't sound fully as the desired behaviour .. although I am not fully sure :-)

Yes, agreed that it's unclear. In that case, I think it's less surprising for `'category'` to behave as `CDT(None, None)`, i.e. don't change anything.

>  So I don't fully understand the comment why it is needed that all CDT(None, ordered=True/False) needs to be equal regardless of the ordered attribute.

That was for backwards comparability. I wouldn't use it as an argument of consistency, since it's a bad thing to be consistent about :)

jschendel committed:
Make CI.astype('category') not change anything

jschendel committed:
Fix tests broken by II.astype('category') changes

jschendel commented:
Updates:
  - Modified `CategoricalIndex.astype('category')` to not change anything
  -  Fixed some tests that previously relied on `IntervalIndex.astype('category')` returning a `Categorical` with `ordered=True` instead of a `CategoricalIndex`
      - This only impacted the creation of `expected` in the tests; didn't change `result`.

jreback commented on pandas/core/indexes/category.py:
```diff
@@ -341,9 +341,26 @@ def __array__(self, dtype=None):
 
     @Appender(_index_shared_docs['astype'])
     def astype(self, dtype, copy=True):
+        if isinstance(dtype, compat.string_types) and dtype == 'category':
```
I don't think you actually need this check as the ``dtype == self.dtype`` line below should pick this up.

jschendel replied:
I think this, or some variation of it, is necessary in order to guarantee that `CI.astype('category')` doesn't change anything.  The issue being that `pandas_dtype('category')` returns `CDT(None, False)`, which would change a `CI` with `ordered=True` to `False`.

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -916,6 +917,10 @@ def astype(self, dtype, copy=True):
                 return self.copy()
             return self
+        elif is_categorical_dtype(dtype):
+            from pandas.core.indexes.category import CategoricalIndex
```
pattern is to import ``CI`` at the top, but ok here too

e.g. ``from pandas.core.indexes.category import CategoricalIndex`` should work

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -632,8 +632,9 @@ def astype(self, dtype, copy=True):
         elif is_categorical_dtype(dtype):
-            from pandas import Categorical
-            return Categorical(self, ordered=True)
+            from pandas.core.indexes.category import CategoricalIndex
```
same

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -2715,9 +2717,14 @@ def difference(self, other):
-                            'is not supported' % self.__class__)
+        dtype = pandas_dtype(dtype)
+        if is_categorical_dtype(dtype):
+            msg = '> 1 ndim Categorical are not supported at this time'
```
test fo this?

jschendel replied:
Wrote a test for it in `test_multi.py`, which overrides the test in `common.py`:

https://github.com/jschendel/pandas/blob/3697d6e49631725a2ecaa91a26f9ae97925b197b/pandas/tests/indexes/test_multi.py#L558-L568

jreback commented on pandas/core/indexes/numeric.py:
```diff
@@ -313,10 +314,14 @@ def astype(self, dtype, copy=True):
         elif is_object_dtype(dtype):
             values = self._values.astype('object', copy=copy)
+        elif is_categorical_dtype(dtype):
+            from pandas.core.indexes.category import CategoricalIndex
```
same

jreback commented on pandas/core/indexes/period.py:
```diff
@@ -517,6 +518,10 @@ def astype(self, dtype, copy=True, how='start'):
             return self.to_timestamp(how=how).tz_localize(dtype.tz)
         elif is_period_dtype(dtype):
             return self.asfreq(freq=dtype.freq)
+        elif is_categorical_dtype(dtype):
```
same

side thing, I *think* that we could make a more generic astype in  ``indexes.base`` and remove *some* boiler plate *maybe* (of course separate PR), you can make an issue if you want (or just PR!)

jreback reviewed (changes requested):
lgtm. small comments.

jschendel committed:
Move CI imports

jschendel commented:
Updated to move the `CI` imports, and replied to the other comments in the latest review.

jorisvandenbossche commented:
Looks good to me

jschendel committed:
refactor dtype update

jschendel commented:
Moved the categorical dtype update code to a new `_update_dtype` method of `CategoricalDtype`, per  https://github.com/pandas-dev/pandas/pull/18710#discussion_r155957618

Not sure if the name/location is appropriate, but can rename/move if need be.  For the time being, I've maintained the existing logic of `.astype('category')` not updating anything.

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -1053,6 +1053,10 @@ def _to_embed(self, keep_tz=False, dtype=None):
     @Appender(_index_shared_docs['astype'])
     def astype(self, dtype, copy=True):
+        if is_categorical_dtype(dtype):
+            from .category import CategoricalIndex
```
I guess we have an import issue if we import this at the top (with the fully qualified path)?

jreback commented:
thanks @jschendel nice patch! keep em coming!
